### PR TITLE
EndTime: fix overflow when doing large comparisons

### LINKED ISSUE #21317: Bug, keep device Alive and send low volume noise broken in Kodi 19
BUG KODI 19

Having consulted the forums it seems confirmed that kodi no longer works to "keep audio devices alive" and "send low volume audio"..  my soundbar times out and switches off regardless of both settings being enabled..

Other users users have confirmed Kodi isn't sending low volume audio for them either..

Running latest Kodi 19 .4.0 on Ubuntu 20.04.4 LTS

I have good normal working sound output via HDMI (using pulse audio from my Caicos HDMI audio (Radeon HD 7450), but even setting it directly to alsa or direct to HDMI Caicos changes nothing..

I'm using ARC to send the sound through TV to my Panasonic soundbar, nothing has changed with my hardware or TV soindbar setup since Kodi 17..

I'm no longer able to keep the soundbar alive with the "Keep audio device alive option" on always, and/or ticking "the send low volume sound" option..

 it used to be fine with kodi 18... Now if I leave a video paused or am in the home screen doors more than about 10 minutes, the soundbar goes into standby...

If I turn the soundbar to max volume when paused I can hear no faint whistle, though maybe it's out of my human ears audible hrtz range?

I've checked my hardware setting in Ubuntu and my pulse logs and the Caicos is still alive and not going to sleep..

Both pulse and ALSA seems to be affected.

Please investigate and fix bug?

### PR DESCRIPTION
**Pull request by lrusak**
This fixes the issues pointed out by @thexai in #22416 

The idea here is to limit the maximum allowable duration of the timer to be `std::chrono::steady_clock::duration::max()` This is typically `std::chrono::nanoseconds::max()` which is about 292 years on the platform I'm working on.

The underlying problem is this
```
std::chrono::milliseconds::max() > std::chrono::nanoseconds::max()
```
This comparison isn't possible without some type of casting. If we cast milliseconds to nanoseconds we overflow. So we could cast nanoseconds to milliseconds (though we may lose precision), my fix is to simply limit the duration to `std::chrono::nanoseconds::max()` (`std::chrono::steady_clock::duration::max()`)as stated above.

This changes the notion of `infinite` in EndTime to be `std::chrono::nanoseconds::max()` and thus we can simplify some methods.

I added a static method `EndTime<>::Max()` which will return the maximum allowable duration that is able to be set in the templated type.

So methods that used `std::chrono::milliseconds::max()` can now use `EndTime<>::Max()`. Using the former won't harm as the duration will simply be truncated in `Set`.

I also added some tests to help with this behavior.

I also added a log message in case we ever see a truncation happen in the future (maybe it should throw instead?) as I believe I fixed the (2) locations that the overflows happened. I'm not sure if anyone noticed the bug in `CGUIListItemLayout` but this fixes the issue there too.

### CODE REVIEW

**Comment by fritsch** [comment]
There is one additional usage of INT_MAX in the code for a timer - I am not at my PC currently, so can't point it out - here care is needed as well.

File: xbmc/threads/SystemClock.h [hunk 1]
```diff
@@ -32,36 +34,39 @@ template<typename T>
...
+
   void Set(const T duration)
   {
     m_startTime = std::chrono::steady_clock::now();
-    m_totalWaitTime = duration;
+
+    if (duration > m_max)
+    {
+      m_totalWaitTime = m_max;
+      CLog::Log(LOGDEBUG, "duration ({}) greater than max ({}) - duration will be truncated!",
```
Comment (thexai) [inline xbmc/threads/SystemClock.h:52]: Better change to LOGWARNING to become always visible

**Comment by thexai** [comment]
It's actually happening:

`2023-01-13 10:15:44.589 T:21096 warning <general>: duration (128849018820000) greater than max (9223372036854) - duration will be truncated!`

Due:

https://github.com/xbmc/xbmc/blob/e9c08df8289b97e539b8b45540ad42de858fffae/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESettings.cpp#L129

This is more than 4000 years!!!

I recommend change this line to:

`std::numeric_limits<short>::max() `

As 546 hours is enough for audio keep alive "Always"

**Comment by fritsch** [comment]
Thanks @thexai - awesome find. Let's make this a WARNING then, then everyone sees it. I would still suggest backporting the proper solution, so that we have a "robust" code-base.

**Review by fritsch — COMMENTED** [review]
Please add the WARNING instead of the DEBUG - then it's good to go I'd say.

**Comment by thexai** [comment]
Seems this is changed to "magic number" that represents ~292 years in minutes. Is also a good solution but then Nexus backport requires all commits in this PR and the previous. Not tested but should work.

File: xbmc/threads/SystemClock.h (hunk 1 above)
Comment (lrusak) [reply xbmc/threads/SystemClock.h]: updated

**Review by fritsch — APPROVED** [review]
Thanks very much. Should go into Nexus in combination with the minutes-fix, then we have a clean / maintainable base there. Awesome.

**Review by thexai — APPROVED** [review]

**Comment by fritsch** [comment]
One feedback from the forum: https://forum.kodi.tv/showthread.php?tid=368027&pid=3131293#pid3131293

**Pull request merged**